File: python/day6.py
from pprint import pprint

def walk(themap, row, col):
    dcol = 0
    drow = -1
    dir = 0
    visted = 1
    seen = {}
    themap[row][col] = 'X'
    while True:
        #pprint(walked)
        if (dir,row,col) in seen:
            return (visted, True)
        seen[(dir,row,col)] = True
        (nextr,nextc) = (row+drow, col+dcol)
        if not ( nextr in range(len(themap)) and nextc in range(len(themap[0])) ):
            break
        if themap[nextr][nextc] == "#":
            dir = (dir + 1) % 4
            match dir:
                case 0 : (dcol,drow) = (0,-1)
                case 1 : (dcol,drow) = (1,0)
                case 2 : (dcol,drow) = (0,1)
                case 3 : (dcol,drow) = (-1,0)
        else:
            if themap[nextr][nextc] != "X":
                visted += 1
                themap[nextr][nextc] = "X"
            col = nextc
            row = nextr

    return (visted, False)

def day6(input):
    themap = []
    col = -1
    for (i, irow) in enumerate(input):
        themap.append(list(irow))
        if '^' in irow:
            row = i
            col = irow.index('^')
    themap[row][col] = '.'
    loops = 0
    mapcopy = [row[:] for row in themap]
    (result1, _) = walk(mapcopy, row, col)
    for orow in range(len(themap)):
        for ocol in range(len(themap[0])):
            if themap[orow][ocol] == ".":
                mapcopy = [l[:] for l in themap]
                mapcopy[orow][ocol] = "#"
                #pprint(mapcopy)
                (_, looped) = walk(mapcopy, row, col)
                if looped:
                    loops += 1
    print(result1,loops)


#41, 6
input = open("src/bin/inputs/day6.test1.txt", "r").readlines()
day6(input)

#4988, 1697
input = open("src/bin/inputs/day6.txt", "r").readlines()
day6(input)

File: src/bin/day6.rs
const INPUT: &str = include_str!("inputs/day6.txt");

pub fn day6(input: &str) -> (usize, usize) {
    (input.len(), input.len())
}

fn main() {
    let (p1, p2) = day6(INPUT);
    println!("{}\n{}", p1, p2);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test1() {
        let input = include_str!("inputs/day6.test1.txt");
        let (p1, p2) = day6(input);

        assert_eq!(p1, input.len());
        assert_eq!(p2, input.len());
    }

    #[test]
    fn test_main() {
        let (p1, p2) = day6(INPUT);

        assert_eq!(p1, INPUT.len());
        assert_eq!(p2, INPUT.len());
    }
}
